### PR TITLE
ENH: add support for molecular fields (AMRex)

### PR DESCRIPTION
**Pull request by simonguichandut**
## PR Summary

The boxlib dataset parsing (for CASTRO/MAESTROeX) are currently set up for usual atomic species like "He4" or "C12", or descriptive names like "ash" with no numbers.  The parsing breaks for e.g. molecules like "H2O" or "C6H6".  This fix escapes those cases.

Ideally, we could have some fancy regex to correctly parse the molecules (and create proper a TeX string), but this is a very niche use of those codes.

## PR Checklist

- [ ] New features are documented, with docstrings and narrative docs
- [ ] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

File: yt/frontends/boxlib/fields.py
Comment (neutrinoceros) [inline yt/frontends/boxlib/fields.py:539]: Bare except statements should be avoided, please only catch specific exceptions. The code in the `try` branch should also be limited to the part that's expected to sometimes raise exceptions.

**Comment by matthewturk** [comment]
Thanks for doing this, @simonguichandut !  It's tough for us to anticipate all the different representations, so making sure we're up to date on how codes represent molecules and aligning that with your expectations is really important.  Thank you!

**Comment by zingale** [comment]
I tested this with a Castro and MAESTROeX plotfile and it works.

File: yt/frontends/boxlib/fields.py
Comment (neutrinoceros) [inline yt/frontends/boxlib/fields.py:534]: We do not use `print` to log errors. You could use `ytLogger.error` here instead.

File: yt/frontends/boxlib/fields.py
Comment (neutrinoceros) [inline yt/frontends/boxlib/fields.py:545]: Given the existence of #4845, where this file is moved, we're going to get merge conflict anyway so let's make their resolution as easy as possible by keeping the diff minimal.
```suggestion
```

File: yt/frontends/boxlib/fields.py
Comment (neutrinoceros) [inline yt/frontends/boxlib/fields.py:499]: I don't think I understand what this branch is supposed to capture. Both examples in the comment (`C12`, `H20`) check `field[-2].isdigit()`, so they both go through, but what I understand from the comment is that `C12` should go, and `H2O` shouldn't. What am I missing ?

File: yt/frontends/boxlib/fields.py
Comment (neutrinoceros) [inline yt/frontends/boxlib/fields.py:533]: Upon a closer look, I don't think we don't need a try/except at all here
```suggestion
    if (spec_match := spec_finder.search(field)) is not None:
        nice_name = "".join(spec_match.groups())
    else:
        # molecules (e.g. H2O, C6H6)
```

**Review by neutrinoceros — COMMENTED** [review]
Thank you for working on this ! Here's a couple questions and suggestions.

File: yt/frontends/boxlib/fields.py
Comment (simonguichandut) [reply yt/frontends/boxlib/fields.py:545]: Just to make sure I understand, you are only suggesting to remove the blank line here?

Comment (zingale) [reply yt/frontends/boxlib/fields.py:545]: I think #4845 is ready to merge, right?
should we wait on anything else in that PR?

File: yt/frontends/boxlib/fields.py
Comment (simonguichandut) [reply yt/frontends/boxlib/fields.py:499]: Actually the strings in the plotfiles are X(C12), X(H2O).  Previously the last character before the ")" parenthesis would be a number, but not in the H2O case, which will raise an error in `weight = int(weight)` a few lines down.  Note that this block of code is for the moment superfluous, element and weight are not used.

File: yt/frontends/boxlib/fields.py
Comment (neutrinoceros) [reply yt/frontends/boxlib/fields.py:545]: > Just to make sure I understand, you are only suggesting to remove the blank line here?

Yes

Comment (neutrinoceros) [reply yt/frontends/boxlib/fields.py:545]: > should we wait on anything else in that PR?

I just want to honor the "one week or so" delay I promised in my last call for objections. Reviews came in faster than I anticipated, but there's no harm in waiting a couple days more, I think.

File: yt/frontends/boxlib/fields.py
Comment (neutrinoceros) [reply yt/frontends/boxlib/fields.py:499]: The comment should reflect that then. Ideally the code would be expressive enough that a comment wouldn't even be needed.
Are you saying the whole block is superfluous ? If so, I think it should be removed, or at least not touched in this PR.

File: yt/frontends/boxlib/fields.py
Comment (simonguichandut) [reply yt/frontends/boxlib/fields.py]: It's not really an "error" and ytLogger prints a whole traceback, not sure that is the ideal solution

File: yt/frontends/boxlib/fields.py
Comment (simonguichandut) [reply yt/frontends/boxlib/fields.py]: I think it was put in there so eventually we could use the element and weight variables (@zingale ?)
https://github.com/yt-project/yt/blob/d7209f9aab2587bca4264f42c2106d4bf6b5cf88/yt/frontends/boxlib/fields.py#L509-L510

Comment (simonguichandut) [reply yt/frontends/boxlib/fields.py]: I only made this change so that the code wouldn't break when there is a species like H2O.

File: yt/frontends/boxlib/fields.py
Comment (neutrinoceros) [reply yt/frontends/boxlib/fields.py]: > It's not really an "error"

please clarify. To me, seeing `return` inside an `except` blocks means we've encountered a recoverable exception (error).


> ytLogger prints a whole traceback

I'm not sure what you mean here. `ytLogger` is meant to print messages to the console and we generally don't include (even partial) tracebacks.

File: yt/frontends/boxlib/fields.py
Comment (neutrinoceros) [reply yt/frontends/boxlib/fields.py]: wouldn't the ideal fix be to extend the regexp so that it doesn't break for these cases ?

File: yt/frontends/boxlib/fields.py
Comment (simonguichandut) [reply yt/frontends/boxlib/fields.py]: Sorry I'm using the wrong terminology. I'm just trying to say that the "error" in this case is not a "big deal". The species is not recognized, so we won't try to produce a nice tex label, that's it.  ytLogger prints out very long error message and the whole call stack, which feels like overkill.

Here's what I get 
```
python $MAESTROEX_DIR/Util/yt/plotsinglevar.py plt_0000100 "X(H2O)"
--- Logging error ---
Traceback (most recent call last):
  File "/Users/simon/anaconda3/lib/python3.11/logging/__init__.py", line 1110, in emit
    msg = self.format(record)
          ^^^^^^^^^^^^^^^^^^^
  File "/Users/simon/anaconda3/lib/python3.11/logging/__init__.py", line 953, in format
    return fmt.format(record)
           ^^^^^^^^^^^^^^^^^^
  File "/Users/simon/anaconda3/lib/python3.11/logging/__init__.py", line 687, in format
    record.message = record.getMessage()
                     ^^^^^^^^^^^^^^^^^^^
  File "/Users/simon/anaconda3/lib/python3.11/logging/__init__.py", line 377, in getMessage
    msg = msg % self.args
          ~~~~^~~~~~~~~~~
TypeError: not all arguments converted during string formatting
Call stack:
  File "/Users/simon/Desktop/codes/AMREX-Astro/MAESTROeX/Util/yt/plotsinglevar.py", line 94, in <module>
    plot_single_var(args.plotfile, args.outfile, args.variables, use_log, args.norm, args.minimum, args.maximum)
  File "/Users/simon/Desktop/codes/AMREX-Astro/MAESTROeX/Util/yt/plotsinglevar.py", line 52, in plot_single_var
    plots = yt.SlicePlot(ds, norm_axis, var_names)
  File "/Users/simon/Desktop/codes/yt-git/yt/visualization/plot_window.py", line 1821, in __init__
    (bounds, center, display_center) = get_window_parameters(
  File "/Users/simon/Desktop/codes/yt-git/yt/visualization/plot_window.py", line 70, in get_window_parameters
    width = ds.coordinates.sanitize_width(axis, width, None)
  File "/Users/simon/Desktop/codes/yt-git/yt/geometry/coordinates/coordinate_handler.py", line 300, in sanitize_width
    self.ds.index
  File "/Users/simon/Desktop/codes/yt-git/yt/data_objects/static_output.py", line 613, in index
    self.create_field_info()
  File "/Users/simon/Desktop/codes/yt-git/yt/data_objects/static_output.py", line 665, in create_field_info
    self.field_info.setup_fluid_fields()
  File "/Users/simon/Desktop/codes/yt-git/yt/frontends/boxlib/fields.py", line 475, in setup_fluid_fields
    nice_name, tex_label = _nice_species_name(field)
  File "/Users/simon/Desktop/codes/yt-git/yt/frontends/boxlib/fields.py", line 534, in _nice_species_name
    ytLogger.error("Could not parse species ", field)
Message: 'Could not parse species '
Arguments: ('X(H2O)',)
```

File: yt/frontends/boxlib/fields.py
Comment (simonguichandut) [reply yt/frontends/boxlib/fields.py]: Yes, I mentioned this when I submitted the PR.  I tried for a while but it was too difficult for me.  Ideally you would also make the numbers subscripts instead of superscript for the tex label for molecules.  Again, having molecules in the first place is a very niche use of castro/maestro...

File: yt/frontends/boxlib/fields.py
Comment (neutrinoceros) [reply yt/frontends/boxlib/fields.py]: The traceback isn't a feature of the logger, but shows a mistake in your call to `ytLogger.error`:

Instead of
```python
ytLogger.error("Could not parse species ", field)
```
write
```python
ytLogger.error("Could not parse species %s", field)
```

File: yt/frontends/boxlib/fields.py
Comment (neutrinoceros) [reply yt/frontends/boxlib/fields.py]: Ok, thanks for clarifying. I suggest we slightly rephrase the logging error message to make it clear that it's a missing feature, not an actual bug, so that readers might choose to give it a try too :)

File: yt/frontends/boxlib/fields.py
Comment (simonguichandut) [reply yt/frontends/boxlib/fields.py]: Oh, thank you!

File: yt/frontends/boxlib/fields.py
Comment (simonguichandut) [reply yt/frontends/boxlib/fields.py]: I wrote a brief something, but please feel free to edit

**Comment by neutrinoceros** [comment]
pre-commit.ci autofix

File: yt/frontends/boxlib/fields.py
Comment (neutrinoceros) [inline yt/frontends/boxlib/fields.py:501]: I still find this too fragile, but digging for a way to make it more robust, I think I ended up solving the problem of parsing molecules. Do you mind if I push to your branch ? I could also open a separate PR.

Comment (simonguichandut) [reply yt/frontends/boxlib/fields.py:501]: Sure, whatever you think is best

Comment (neutrinoceros) [reply yt/frontends/boxlib/fields.py:501]: ok I'll set myself a reminder to execute this by next week, thank you for your patience !

**Comment by neutrinoceros** [comment]
I'm going to follow through my comment from last week now. First, I'll rebase the branch to resolve merge conflicts.

**Comment by neutrinoceros** [comment]
Here goes nothing

File: yt/frontends/amrex/fields.py
Comment (neutrinoceros) [inline yt/frontends/amrex/fields.py:506]: As I was refactoring this branch, I realized it had really been left in a noop "TODO" state for 10 years and I feel like it's not worth keeping it around given that it creates discussion around code that's actually not used.

**Comment by neutrinoceros** [comment]
I ended up radically re-orienting the PR's implementation. Hopefully the behavior I implemented is suitable.

File: yt/frontends/amrex/fields.py
Comment (yut23) [reply yt/frontends/amrex/fields.py:506]: We should also remove the same code in `CastroFieldInfo.setup_fluid_fields()`.

File: nose_ignores.txt
Comment (neutrinoceros) [reply nose_ignores.txt:46]: indeed I forgot about this. Thank you !

**Comment by neutrinoceros** [comment]
Ok, CI is green now, so let's open back for review. @simonguichandut what do you think ?

File: yt/frontends/amrex/fields.py
Comment (neutrinoceros) [reply yt/frontends/amrex/fields.py:506]: Good point. Actually a separable problem so I'll open another PR. Thank you !

**Comment by yut23** [comment]
The TeX formatting is a bit wonky now: molecules and descriptive names are missing the "X" that the isotopes have, and omega dot for isotopes is rendered like `\dot{\omega}[X(He4)]`. I think it would be better to have `Substance.to_tex()` return just the rendered name (e.g. `^{4}He` and `C_{12}H_{24}`) and have `setup_fluid_fields()` add the parentheses and `X`/`\dot{\omega}`.

**Comment by neutrinoceros** [comment]
Good points ! Separation of concerns FTW ! better now ?

**Comment by yut23** [comment]
Yes, much better. Can we also add the tex formatting to CastroFieldInfo?

**Comment by neutrinoceros** [comment]
done !
I'll rebase the branch again if you're happy with it

File: yt/frontends/amrex/fields.py
Comment (yut23) [inline yt/frontends/amrex/fields.py:372]: To match Maestro:
```suggestion
                    display_name=rf"\rho {sub.to_tex()}",
```

File: yt/frontends/amrex/fields.py
Comment (yut23) [inline yt/frontends/amrex/fields.py:364]:
```suggestion
                sub = Substance(field)
                # Overwrite field to use nicer tex_label display_name
                self.add_output_field(
                    ("boxlib", field),
                    sampling_type="cell",
                    units="",
                    display_name=rf"X\left({sub.to_tex()}\right)",
                )
```

**Review by yut23 — APPROVED** [review]

**Comment by neutrinoceros** [comment]
thanks @yut23 for your input ! Now I think we should give @simonguichandut a couple weeks to comments before we can merge.

File: yt/frontends/amrex/fields.py
Comment (neutrinoceros) [reply yt/frontends/amrex/fields.py]: nice catch. I adjusted tests and fixed this bug as you suggested !

**Comment by simonguichandut** [comment]
Looks great to me! I've tested on MAESTROeX plotfiles, with and without molecules, and it works as expected.

**Comment by neutrinoceros** [comment]
Awesome, I squashed all my commits together, let me know if you're okay with doing the same with yours !

**Comment by simonguichandut** [comment]
> Awesome, I squashed all my commits together, let me know if I should do the same with yours !

I'm not sure what this means/does, but either way I will delete my branch and sync with the main repo! :)

**Comment by neutrinoceros** [comment]
I'm proposing to rewrite the branch history so  it contains just one commit. With some tools (mainly `git bisect`) makes it more confortable to debug if we need to revisit this change in the future. The reason I asked is that I don't want to make your life harder in any way, but if it sounds like it wouldn't, so I'll just go ahead.

**Review by yut23 — APPROVED** [review]
Looks good to me!

**Comment by zingale** [comment]
How does this handle something like "H2", with is either an isotope (deuterium) or a molecule (molecular H)?
We probably want to default to isotopes.

**Comment by simonguichandut** [comment]
> How does this handle something like "H2", with is either an isotope (deuterium) or a molecule (molecular H)? We probably want to default to isotopes.

H2 does get recognized as the isotope

**Comment by neutrinoceros** [comment]
Yes, that's intentional, because from what I understand molecules are rarely seen in this context, but it's a shame that there's no way to tell for sure from just looking at the string `"H2"`. 
@zingale I could sediment the current behavior by adding a test case for it if you'd like.

**Comment by zingale** [comment]
no, I think it's fine.  Just wanted to double check.

**Comment by neutrinoceros** [comment]
Alright let's just merge then. Thanks all !

**Pull request merged**
